src/Confluent.Kafka/Producer.cs

```diff
@@ -687,8 +687,8 @@ internal Producer(ProducerBuilder<TKey, TValue> builder)
             try
             {
                 valBytes = (valueSerializer != null)
-                    ? valueSerializer.Serialize(message.Value, false, message, topicPartition)
-                    : asyncValueSerializer.SerializeAsync(message.Value, false, message, topicPartition)
+                    ? valueSerializer.Serialize(message.Value, new SerializationContext(MessageComponentType.Value, topicPartition.Topic))
```

Comment:
I don't know. a couple of things:
1. I recall considering making this a struct (which means this is stack allocated, and so I think would result in little overhead c.f. the parameters themselves), but for some reason decided on class. I can't remember why. I'll think about that again.
2. With the builder class, we could add overloads to `SetKeySerializer` etc. in the future to accept a `ISimpleSerializer` or something which doesn't have this argument (if that results in a significant perf improvement). These could be used for default serializers as well. All non-breaking changes.

Reply:
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/choosing-between-class-and-struct

It fails the first test (meaning the guidelines call for a class), but not the others. My feeling is a struct will be better. I've made that change in the commit i just added, but I won't merge until after benchmarking.

Reply:
I attempted to measure the difference between class and struct - the results are pretty noisy and I can't tell the difference. Going with `struct`, as I believe it's theoretically a bit more performant. note: it won't be a breaking change in the future if for some reason we want to revert (say we get many more items in `SerializationContext`.

Reply:
now i see that all the tests failing :-), I remember why I made this a class - because it's nullable. This is convenient for using the serdes outside the context of the producer/consumer because there's often no useful context / this is usually irrelevant to the serde. However, that thinking is backwards. The clients always pass in a non-null reference, so it is in fact desirable that the interface reflects that.